qiskit/backends/ibmq/ibmqjob.py

```diff
@@ -359,86 +312,91 @@ def _wait_for_job(self, timeout=60, wait=5):
-        job_result = {'job_id': job_id,
-                      'status': api_result['status'],
-                      'used_credits': api_result.get('usedCredits'),
-                      'result': job_result_list}
-        # logger.info('Got a result for qobj: %s from remote backend %s with job id: %s',
-        #             qobj["id"], qobj['config']['backend_name'],
-        #             job_id)
-        job_result['backend_name'] = self.backend_name
-        return Result(job_result)
+                raise IBMQJobError("get_job didn't return status: %s" %
+                                   (pprint.pformat(api_result)))
+
+            if 'ERROR' in api_result['status'].upper():
+                logger.info('There was an error in the Job: status = %s', api_result['status'])
+                self._status = JobStatus.ERROR
+                job_result = JobResult(self._job_id, 'ERROR', api_result['status'])
+                break
+
+            if api_result['status'] == 'RUNNING':
+                logger.info('status = %s (%d seconds)', api_result['status'], timer)
+                time.sleep(wait)
+                timer += wait
+                continue
+
+            if api_result['status'] == 'CANCELLED':
+                self._status = JobStatus.CANCELLED
+                job_result = JobResult(self._job_id, 'CANCELLED', api_result['status'])
+                logger.info('The job was cancelled!')
+                break
+
+            # After this point the job is completed!
+            logger.info('Job is complete! status = %s', api_result['status'])
+            job_result_list = []
+            for circuit_result in api_result['qasms']:
+                this_result = {'data': circuit_result['data'],
+                               'name': circuit_result.get('name'),
+                               'compiled_circuit_qasm': circuit_result.get('qasm'),
+                               'status': circuit_result['status']}
+                if 'metadata' in circuit_result:
+                    this_result['metadata'] = circuit_result['metadata']
+                job_result_list.append(this_result)
+
+            logger.info('Got a result from remote backend %s with job id: %s',
+                        self.backend_name, self._job_id)
+            job_result = JobResult(self._job_id, api_result['status'],
+                                   job_result_list,
+                                   api_result.get('usedCredits'),
+                                   backend_name=self._backend_name)
+            self._status = JobStatus.DONE
+            break
 
+        return Result(job_result.as_dict())
 
-class IBMQJobError(QISKitError):
-    """class for IBM Q Job errors"""
-    pass
+    def _is_commercial(self):
+        config = self._api.config
+        # this check may give false positives so should probably be improved
+        return config.get('hub') and config.get('group') and config.get('project')
+
+    def _wait_for_submitting(self):
```

Comment:
This method requires one to call this instead of checking the status of self but it looks like that means less api calls which sounds good!

Reply:
... and it deals with the problem of not having a job_id if a public method is called, avoiding the CPU hungry loops with sleep(0.1) :)